### ISSUE TITLE
Data leakage from the future?

### ISSUE THREAD

**Issue by erickmiller** [zshicode/MambaStock#3]
Impressive results!   Predictions this good are rare.  

I'm not as familiar with the Mamba architecture am just starting to research this, and found your code and tested it, with this incredibly accurate better than state-of-the-art result, surpassing the prediction power of any other model I have seen:

<img width="1428" alt="image" src="https://github.com/zshicode/MambaStock/assets/2259464/524d594a-8660-4f98-851b-ded04e4434e2">

Are you sure there's not some test data leakage?  Like data from the future leaking into the predicted results?  

Mamba and your code implementation looks very promising -- but I am confused about why you're not using some type of **time masking** or **temporal splitting** on the test data when you send into `PredictWithData` in order to avoid or eliminate the model having look-ahead bias and/or some sort of data leaking in from the future -- my theory is there is future-data leaking into the predictions, but allow me to explain why -- and please correct me and tell me I am wrong if this is not the case.  It seems that the model should be pre-trained, and then should be doing it's predictions in a loop, and each time the prediction is run, Mamba should only be given the data up until the day-of or the day-before the prediction, and it should not be given all of the past, current and future data to run it's predictions.

Your code appears concise and clear, which makes it easy to read.  But, let me make sure I'm understanding what's going on and that I don't misunderstand:
  
It looks like you're removing the `'close', 'pre_close', 'change', 'pct_chg'` data, and then using: `open,high,low,pre_close,change,pct_chg,vol,amount,turnover_rate,volume_ratio,pe,pb,ps,total_share,float_share,free_share,total_mv,circ_mv`  to train the model on the closing percent change? 

But you also pass in the entire test set of all future test data (except without the `'close', 'pre_close', 'change', 'pct_chg'` data) into `PredictWithData`.  You then use this in the prediction code after training to predict `yhat`.

In the `PredictWithData` function, first you initialize (and train) the Mamba network using pytorch:
```
clf = Net(len(trainX[0]),1)
```

Before training you load up the test set into the `xv` variable:
```
xv = torch.from_numpy(testX).float().unsqueeze(0)
```

After training you evaluate and predict by passing in the entire set of testX data.
```
clf.eval()
mat = clf(xv)
yhat = mat.detach().numpy().flatten()
```

From what I can see, `xv` contains the entire test set of `testX` that is passed into the `PredictWithData` function. This test set includes data from future time points relative to the training data. By passing the entire test set into the trained model `clf` during the evaluation phase, it appears that the model has access to future information when making predictions.

My concern and/or hypothesis is that this approach introduces a form of data leakage or look-ahead bias, where the model is inadvertently using information from the future to make predictions. This could lead to overly optimistic results that may not generalize well to real-world scenarios where future data is not available.

You do remove the `'close', 'pre_close', 'change', 'pct_chg'` columns from the test set before passing it into `PredictWithData`, but the model still has access to other future information such as `'open', 'high', 'low', 'vol', 'amount', [...]` etc.

I was wondering if you have considered applying some form of **time masking** or **temporal splitting** to ensure that the model only uses information available up to the current time step when making predictions for future time steps. This would help mitigate the risk of data leakage from the future and provide a more realistic evaluation of the model's performance.

I'm relatively new to the Mamba architecture and would greatly appreciate your insights on this matter. Could you please clarify whether there is indeed a potential issue with data leakage in the current implementation and, if so, suggest any modifications or best practices to address it?   I'd really like to test applying this model to the problem of time series classification but in the current state, I believe that the lack of time masking presents the likelihood that data leakage from the future is the reason the results are so incredibly good. 

Read more about time masking and the importance of this in time sequence / time series modeling and training here: 
https://wandb.ai/mostafaibrahim17/ml-articles/reports/A-Deep-Dive-Into-Time-Masking-Using-PyTorch--Vmlldzo1Njg1Nzc5#understanding-time-masking-techniques

**Comment by zshicode** [comment]
Thanks for your detailed comments. However, in my opinion, the code has avoided data leakage to ensure the validity of the results.

(1) The entire data are first divided into training set and testing set. Then, the model adopts the given features to predict the target by minimizing the loss in training set, i.e. the error between target and prediction in training set. **The testing set is not adopted for minimizing the loss function of the model, ensureing that there is no data leakage during the training procedure**.

(2) In evaluating procedure, although the samples in testing set are put into the model simultaneously, Mamba model only adopts features in previous timestamps for prediction. **The future features are not adopted for predicting the target, ensureing that there is no data leakage during the evaluating procedure**.

**Comment by erickmiller** [comment]
I agree with your assessment.  Independent of your note about the Mamba architecture's adoption of features from the previous time steps during prediction -- I decided to do an empirical test where I very mildly refactor the code (making it a little bit less efficient but more separated and clear for my test) so train and predict are separate functions, and then force-truncated the future data from the un-seen test data in a python loop inefficiently running predict each time, and only collecting the last data point of the prediction.  I also offset the % change  for the `trainY` data by one day so that the prior day's data is trained to predict the next day's % change.   The results of my experiment seem to point towards an empirically deduced conclusion that is in-line with your more direct explanation of the Mamba architecture. 

First I renamed the Net class that's inheriting from the torch neural network (this was not necessary but just for my own initial mental clarity):

```python

class MambaNet(nn.Module):
    def __init__(self,in_dim,out_dim):
        super().__init__()
        self.config = MambaConfig(d_model=args.hidden, n_layers=args.layer)
        self.mamba = nn.Sequential(
            nn.Linear(in_dim,args.hidden),
            Mamba(self.config),
            nn.Linear(args.hidden,out_dim),
            nn.Tanh()
        )
    
    def forward(self,x):
        x = self.mamba(x)
        return x.flatten()

```

Next, I mildly refactored the code into train and predict code, with a new class, so training and prediction could happen independently and state could be saved in memory of the object after training.  I was going to implement saving a checkpoint with early stopping but that seemed outside the scope of the test and training is SO FAST :) so  didn't really need to in order to test my  hypothesis.  Here's the test code refactored into a class - slightly less efficient but from a practical perspective, nice to be able to call predict independent of train:

```python

class MambaStock():
    def __init__(self, trainX):
        super().__init__()
        self.mambaNetwork = MambaNet(len(trainX[0]),1)
        self.trainX = trainX

    def train(self, trainy):
        opt = torch.optim.Adam(self.mambaNetwork.parameters(),lr=args.lr,weight_decay=args.wd)
        xt = torch.from_numpy(self.trainX).float().unsqueeze(0)
        yt = torch.from_numpy(trainy).float()
        if args.cuda:
            self.mambaNetwork = self.mambaNetwork.cuda()
            xt = xt.cuda()
            yt = yt.cuda()
        for e in range(args.epochs):
            self.mambaNetwork.train()
            z = self.mambaNetwork(xt)
            loss = F.mse_loss(z,yt)
            opt.zero_grad()
            loss.backward()
            opt.step()
            if e%10 == 0 and e!=0:
                print('Epoch %d | Lossp: %.4f' % (e, loss.item()))
        self.mambaNetwork.eval()

    def predict(self, testX):
        xv = torch.from_numpy(testX).float().unsqueeze(0)
        if args.cuda:
            self.mambaNetwork = self.mambaNetwork.cuda()
            xv = xv.cuda()
        self.mambaNetwork.eval()
        mat = self.mambaNetwork(xv)
        if args.cuda: mat = mat.cpu()
        yhat = mat.detach().numpy().flatten()
        return yhat


```

Next, again just for my own mental clarity, unnecessarily put  the `n_test` argument into a couple variables, `split` and `split_point` since this is the point where the data gets split into train (seen) and test (unseen) data.  This is also where I offset the % change one step into the future so the model is trained to predict tomorrow's % change rather than the current day's % change.  My intuition (and assumption) was that the data `vol, amount, turnover_rate, volume_ratio, pe, pb, ps, total_share, float_share, free_share, total_mv, circ_mv` is all data measured at market close?  If so, then this data would not be available on the day the prediction needs to be made, which is the main reason I offset the % change.   

[Note: I tried to first gain more insight into the nature of the data measurements by making an account on the website you provided to get more information about when these data points are being measured, but I think the website is not working for me or something, the text message confirmation to allow me to create an account and pass through auth never arrived at my phone number.]

In any case, either way, offsetting the % change just further adds one degree of distance from the present time into the future time during training so figured either way, it would work for the test:

```python

split = args.n_test
split_point = -split

data = pd.read_csv(args.ts_code+'.SH.csv')
data['trade_date'] = pd.to_datetime(data['trade_date'], format='%Y%m%d')
close = data.pop('close').values
ratechg = data['pct_chg'].apply(lambda x:0.01*x).values
data.drop(columns=['pre_close','change','pct_chg'],inplace=True)
dat = data.iloc[:,2:].values

trainX = dat[:split_point-1, :]  # Truncate the last element to align sizes
trainY = ratechg[1:split_point]   # Start from 1 to offset by one step forward
mamba = MambaStock(trainX)
mamba.train(trainY)

testX =  dat[split_point-2:, :]

```

Finally, the really unnecessary part :) where I call predict in an inefficient loop over and over again.  This is the truly empirical part of the test, rather than dissect the Mamba architecture and this implementation, I figured "lets see what happens":

```python

# predictions = mamba.predict(testX)

predictions = []
for i in range(1, split+1):
    # trainX[i:, :]
    prediction = mamba.predict(testX[:i, :])
    predictions.append( prediction[-1] )

time = data['trade_date'][split_point:]
data1 = close[split_point:]
finalpredicted_stock_price = []
pred = close[split_point-1]
for i in range(split):
    pred = close[split_point-1+i]*(1+predictions[i])
    finalpredicted_stock_price.append(pred)

```

And the result was pretty good.  Consistent with your conclusion I think.  

<img width="351" alt="image" src="https://github.com/zshicode/MambaStock/assets/2259464/8e8f2850-9930-4ed3-9316-af6c299b37fd">

The offsetting in time seems to have skewed the predictions forward, and maybe I made an error in the way I offset the data relative to the current date, as I wrote this code really quickly last night, so I'm going to look at that again today to double check.  But overall pretty cool.  The simplicity of the Mamba implementation and the speed of training and inference in this relatively low-dimensional situation (relative to the high-dimensionality of a typical transformer) is really cool.  I'm going to have to dive in a bit deeper to the state space model.   

<img width="1581" alt="image" src="https://github.com/zshicode/MambaStock/assets/2259464/a70f8505-30a5-4762-980e-31094c2051b5">

Cool.  I'm very curious to see how this architecture performs on a broader classification task where it's trained on more than one dimension data that is more like a one-hot encoding [[1,1,0,0], [0,0,0,0], [0,0,0,0]] and on data that is less clearly and/or directly immediately correlated with the input data.  

One thing I was thinking/wondering about, not sure if you could point me towards a paper or if the Mamba architecture already solves this problem - there is a semi-recurrent and/or dependent nature to many classification tasks, where the prior predictions/classifications form a pattern on their own that should inform the next prediction with a much stronger accuracy.  

This at first appears like it could be non-deterministic but I think is simple to unroll into an n-graph-like decision layer, or maybe just a convolutional layer or something -- such that, for example in a 3 label classification task, for labels `-1, 0, 1` when the model is trained to predict `1`,  the prior sequence of predictions or actual classifications should be such that it only should predict `1` if all of the past predictions were either `0` and/or `-1` -- and the same with `0` or `-1` conversely (it should only predict `-1` if the prior history included all `0` as well as positive `1`) -- such that the prior states and/or pattern of the value to predict and/or the predicted values help to create causality to inform the accuracy and relevancy for predictions of the next value.  As any problem of recurrence, this can be unrolled into a logic tree, sort of like a decision layer.   

LSTM and RNNs or WaveNets come to mind, but even in all these cases, my understanding is the recurrence doesn't consider the priors of the predicted values, only the priors of the training data.   When I look at things like self-attention or cross-attention, etc, all I'm seeing is memory of the input data but not learning the pattern of the past predictions themselves over-time to better choose the next prediction -- this type of learning seems particularly important when the classifications represent or are tied to inter-dependent past decisions -- as are are many (maybe most) temporal based classification tasks.  The ability to do this in a separate training phase with a reinforcement learning environment came to mind, but even if this was the case, it seems that a pre-trained model should still be informed of the history of it's own decisions during training, when the past decisions influence the future decisions.

Do you know of architectures or an implementation of a paper or a network that does this -- combined with the look-back prediction power of a modern transformer w/ attention or a mamba state-space like architecture?

It seems like any data that occurs across time would benefit from training an additional network layer -- to simply learn from the pattern of it's past decisions, to better inform it's next future decision.

**Comment by erickmiller** [comment]
After writing this, I started researching more and quickly found some cool architectures for transformers with feedback loops, then found this -- which looks super cool:  https://github.com/Toshihiro-Ota/decision-mamba 

This was off topic and is not related to any issue in your repo, but figured I'd post my quick findings here for posterity.

**Comment by brunodifranco** [comment]
First of all, great work. You're results are amazing and the code is easy to reproduce and read. I'm new to the Mamba architecture, and still have a few questions :D.

 I'm aware this issue has been close, but i'm really curious regarding the following comment:

"Mamba model only adopts features in previous timestamps for prediction."

Can you point in the Mamba code where this is shown?

This means that if the model is predicting tomorrows stock return, it's using today's open, low, high and volume features for instance?
